lgc/state/PipelineState.cpp

```diff
@@ -524,11 +524,26 @@ void PipelineState::setUserDataNodesTable(ArrayRef<ResourceNode> nodes, Resource
 
     // Copy the node.
     destNode = node;
-    if (node.type == ResourceNodeType::DescriptorTableVaPtr) {
+
+    switch (node.type) {
+    case ResourceNodeType::DescriptorTableVaPtr:
       // Process an inner table.
       destInnerTable -= node.innerTable.size();
       destNode.innerTable = ArrayRef<ResourceNode>(destInnerTable, node.innerTable.size());
       setUserDataNodesTable(node.innerTable, destInnerTable, destInnerTable);
+      break;
+    case ResourceNodeType::IndirectUserDataVaPtr:
+    case ResourceNodeType::StreamOutTableVaPtr:
+      break;
+    default:
+      // If there is immutable sampler data, take our own copy of it.
+      if (node.immutableSize != 0) {
+        unsigned sizeInDwords = node.immutableSize * DescriptorSizeSampler;
+        m_immutableValueAllocs.push_back(std::make_unique<uint32_t[]>(sizeInDwords));
+        memcpy(&m_immutableValueAllocs.back()[0], node.immutableValue, sizeInDwords);
```

Comment:
Shouldn't this use the size in bytes instead of dwords? Consider using `llvm::copy`/`std::copy` with operator on ranges/iterators

Reply:
Oops, yes, thanks. Actually "sizeInDwords" is the size in bytes, because DescriptorSizeSampler is the size of a sampler descriptor in bytes. So the code as written works, but it has a confusing variable name and it allocates 4x too much for the buffer. I have resolved this by really making sizeInDwords the size in dwords, and using std::copy as you suggest. Also fixed a similar problem in a comment in Pipeline.h.